pkg/logutil/merge_logger.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logutil
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/coreos/pkg/capnslog"
+)
+
+// line represents a log line that can be printed out
+// through capnslog.PackageLogger.
+type line struct {
+	level capnslog.LogLevel
+	str   string
+}
+
+type status struct {
+	gap        time.Duration
+	printTimer *time.Timer
+
+	mu    sync.Mutex // protect fields below
+	start time.Time  // time to start merging
+	count int        // number of lines from start field
+}
+
+// MergeLogger merges repeated log lines and prints summary log lines.
+// It doesn't merge log at DEBUG and TRACE level because they should
+// be verbosal.
+//
+// For one log line, MergeLogger prints out the line when the line appears
+// at the first time. Then it holds and merges lines with the same
```

Comment:
The mergeLogger holds the same log line printed within the same `Gap`. For example...